src/java.base/share/classes/jdk/internal/constant/IndexedComputedConstantMap.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.constant;
+
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.annotation.Stable;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.ToIntFunction;
+
+public final class IndexedComputedConstantMap<K, V> extends AbstractMap<K, ComputedConstant<V>> {
+
+    private static final long ARRAY_BASE_OFFSET = Unsafe.getUnsafe().arrayBaseOffset(ComputedConstant[].class);
+    private static final long ARRAY_INDEX_SCALE = Unsafe.getUnsafe().arrayIndexScale(ComputedConstant[].class);
+
+    private final IntFunction<? extends K> keyMapper;
+    private final ToIntFunction<? super K> keyUnmapper;
+    private final Function<? super K, ? extends V> provider;
+    @Stable
+    private final ComputedConstant<V>[] values;
+
+    @SuppressWarnings("unchecked")
+    public IndexedComputedConstantMap(int size,
```

Comment:
In theory, it would be possible to assert the invariant that the keyMapper and keyUnmapper match. However, this might defeat the purpose of the construct.